File: ForgeHub/server/Api/LogsApi.cs
using System.Text.Json;
using ForgeHub.Data;

namespace ForgeHub.Api;

public static class LogsApi
{
    /// <summary>Berapa banyak baris yang diterima dalam satu kiriman.</summary>
    private const int MaxLinesPerBatch = 1000;

    public static void MapLogs(this WebApplication app)
    {
        // --------------------------------------------------------------
        // Penerimaan log dari robot. Inilah yang dipanggil JakRunner tiap 3 detik.
        //
        // Satu bundel berisi banyak baris dan ditulis dalam SATU transaksi.
        // Seribu penulisan terpisah ke SQLite berarti seribu penyelarasan ke
        // cakram; satu transaksi menjadikannya satu.
        app.MapPost("/api/logs", async (HttpContext context) =>
        {
            var body = await context.Body();

            if (body.ValueKind != JsonValueKind.Object ||
                !body.TryGetProperty("lines", out var lines) ||
                lines.ValueKind != JsonValueKind.Array)
                return Results.BadRequest(new { error = "Butuh { \"lines\": [ ... ] }." });

            if (lines.GetArrayLength() > MaxLinesPerBatch)
                return Results.BadRequest(new
                {
                    error = "Terlalu banyak baris dalam satu kiriman. Batasnya " + MaxLinesPerBatch + ".",
                });

            using var connection = context.Db(out var tenantId);
            using var transaction = connection.BeginTransaction();

            var written = 0;

            foreach (var line in lines.EnumerateArray())
            {
                var message = line.Str("message");
                if (string.IsNullOrWhiteSpace(message)) continue;

                var level = (line.Str("level", "INFO") ?? "INFO").ToUpperInvariant();
                if (level is not ("TRACE" or "DEBUG" or "INFO" or "WARN" or "WARNING" or "ERROR" or "FATAL"))
                    level = "INFO";

                Sql.Exec(connection,
                    @"INSERT INTO logs (tenant_id, level, message, robot_name, machine_name, process_name, job_id, logged_at)
                      VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7)",
                    tenantId, level, message, line.Str("robotName"), line.Str("machineName"),
                    line.Str("processName"), line.Str("jobId"), line.Str("loggedAt", Sql.Now()));

                written++;

                // Kesalahan dari robot juga menjadi peringatan. Log dibaca kalau
                // ada yang sengaja mencarinya; peringatan muncul sendiri.
                if (level is "ERROR" or "FATAL")
                {
                    ApiSupport.Alert(connection, tenantId, "Error",
                        "Kesalahan pada " + (line.Str("robotName") ?? "robot"),
                        message, "logs");
                }
            }

            transaction.Commit();

            return Results.Ok(new { ok = true, written });
        });

        // --------------------------------------------------------------
        // Pembacaan log untuk panel Real-Time Logs.
        //
        // Penjemputannya berdasar afterId, bukan waktu. Stempel waktu datang
        // dari jam robot, dan jam yang meleset beberapa detik membuat baris baru
        // tampak lebih tua daripada yang sudah tampil — lalu tidak pernah muncul.
        // Nomor urut basis data selalu naik apa pun jam si pengirim.
        app.MapGet("/api/logs", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var limit = context.IntQuery("limit", 100);
            var level = context.Request.Query["level"].ToString();
            var robot = context.Request.Query["robot"].ToString();
            var search = context.Request.Query["q"].ToString();

            long afterId = 0;
            long.TryParse(context.Request.Query["afterId"].ToString(), out afterId);

            var sql = @"SELECT id, level, message, robot_name, machine_name, process_name, job_id, logged_at
                          FROM logs
                         WHERE tenant_id = @p0 AND id > @p1";

            var args = new List<object> { tenantId, afterId };

            if (!string.IsNullOrEmpty(level))
            {
                sql += " AND level = @p" + args.Count;
                args.Add(level.ToUpperInvariant());
            }

            if (!string.IsNullOrEmpty(robot))
            {
                sql += " AND robot_name = @p" + args.Count;
                args.Add(robot);
            }

            // Penyaringan per PEKERJAAN: inilah yang membuat panel "log robot
            // yang sedang berjalan" hanya memuat baris dari jalan itu, bukan
            // seluruh riwayat penyewa yang tercampur.
            var job = context.Request.Query["jobId"].ToString();

            if (!string.IsNullOrEmpty(job))
            {
                sql += " AND job_id = @p" + args.Count;
                args.Add(job);
            }

            var process = context.Request.Query["process"].ToString();

            if (!string.IsNullOrEmpty(process))
            {
                sql += " AND process_name = @p" + args.Count;
                args.Add(process);
            }

            if (!string.IsNullOrEmpty(search))
            {
                sql += " AND message LIKE @p" + args.Count;
                args.Add("%" + search + "%");
            }

            // Yang TERBARU diambil lebih dulu lalu dibalik urutannya, supaya
            // permintaan pertama (afterId = 0) memberi 100 baris terakhir,
            // bukan 100 baris pertama dari hari pemasangan.
            sql += " ORDER BY id DESC LIMIT @p" + args.Count;
            args.Add(limit);

            var rows = Sql.Rows(connection, sql, args.ToArray());
            rows.Reverse();

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        app.MapDelete("/api/logs", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var removed = Sql.Exec(connection, "DELETE FROM logs WHERE tenant_id = @p0", tenantId);

            return Results.Ok(new { ok = true, removed });
        });

        MapAlerts(app);
    }

    // ------------------------------------------------------------------

    private static void MapAlerts(WebApplication app)
    {
        app.MapGet("/api/alerts", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var limit = context.IntQuery("limit", 50);
            var unreadOnly = context.Request.Query["unread"].ToString() == "1";

            var rows = unreadOnly
                ? Sql.Rows(connection,
                    @"SELECT id, severity, title, message, source, is_read, created_at
                        FROM alerts WHERE tenant_id = @p0 AND is_read = 0
                       ORDER BY id DESC LIMIT @p1", tenantId, limit)
                : Sql.Rows(connection,
                    @"SELECT id, severity, title, message, source, is_read, created_at
                        FROM alerts WHERE tenant_id = @p0
                       ORDER BY id DESC LIMIT @p1", tenantId, limit);

            return Results.Ok(rows);
        });

        app.MapPost("/api/alerts/{id:long}/read", (HttpContext context, long id) =>
        {
            using var connection = context.Db(out var tenantId);

            Sql.Exec(connection, "UPDATE alerts SET is_read = 1 WHERE tenant_id = @p0 AND id = @p1", tenantId, id);

            return Results.Ok(new { ok = true });
        });

        app.MapPost("/api/alerts/read-all", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var changed = Sql.Exec(connection,
                "UPDATE alerts SET is_read = 1 WHERE tenant_id = @p0 AND is_read = 0", tenantId);

            return Results.Ok(new { ok = true, changed });
        });
    }
}

File: ForgeHub/backend/src/main/java/id/jakforge/forgehub/api/LogsApi.java
package id.jakforge.forgehub.api;

import id.jakforge.forgehub.data.Db;
import id.jakforge.forgehub.security.CurrentUser;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * Catatan dari robot.
 *
 * <p>Dikirim BERKELOMPOK, bukan satu per satu. Robot menghasilkan puluhan
 * baris per detik saat berjalan, dan satu permintaan HTTP per baris berarti
 * robotnya menghabiskan lebih banyak waktu menunggu jaringan daripada
 * mengerjakan pekerjaannya.
 */
@RestController
@RequestMapping("/api/logs")
public class LogsApi {

    /**
     * Batas baris per kiriman.
     *
     * <p>Ada supaya satu kiriman tidak bisa menahan transaksi terlalu lama dan
     * menghalangi dasbor membaca. Robot yang punya lebih banyak akan
     * mengirimkannya sebagai kelompok berikutnya.
     */
    private static final int MAKS_BARIS = 1000;

    private static final Set<String> TINGKAT_SAH =
            Set.of("TRACE", "DEBUG", "INFO", "WARN", "WARNING", "ERROR", "FATAL");

    private final Db db;

    public LogsApi(Db db) {
        this.db = db;
    }

    @GetMapping
    public List<Map<String, Object>> daftar(
            @RequestParam(required = false) String level,
            @RequestParam(required = false) String robot,
            @RequestParam(required = false) String process,
            @RequestParam(name = "jobId", required = false) String jobId,
            @RequestParam(required = false) Integer limit) {

        List<String> where = new ArrayList<>();
        List<Object> args = new ArrayList<>();

        where.add("tenant_id = ?");
        args.add(CurrentUser.get().tenantId());

        if (level != null && !level.isBlank()) {
            where.add("level = ?");
            args.add(level.toUpperCase(Locale.ROOT));
        }

        if (robot != null && !robot.isBlank()) {
            where.add("robot_name = ?");
            args.add(robot);
        }

        if (process != null && !process.isBlank()) {
            where.add("process_name = ?");
            args.add(process);
        }

        if (jobId != null && !jobId.isBlank()) {
            UUID id = Db.uuid(jobId);

            // Id yang bukan UUID tidak akan pernah cocok dengan apa pun. Yang
            // dikembalikan daftar kosong, bukan 500 karena kegagalan penguraian
            // di dalam basis data.
            if (id == null) return Db.kosong();

            where.add("job_id = ?");
            args.add(id);
        }

        args.add(Badan.batas(limit, 200, 2000));

        return db.rows("""
                SELECT id, level, message, robot_name, machine_name, process_name, job_id, logged_at
                  FROM logs
                 WHERE %s
                 ORDER BY id DESC
                 LIMIT ?
                """.formatted(String.join(" AND ", where)), args.toArray());
    }

    @PostMapping
    @Transactional
    @SuppressWarnings("unchecked")
    public ResponseEntity<?> tulis(@RequestBody(required = false) Map<String, Object> body) {
        Object mentah = body == null ? null : body.get("lines");

        if (!(mentah instanceof List<?> lines)) {
            return ResponseEntity.badRequest().body(Map.of("error", "Butuh { \"lines\": [ ... ] }."));
        }

        if (lines.size() > MAKS_BARIS) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Terlalu banyak baris dalam satu kiriman. Batasnya " + MAKS_BARIS + "."));
        }

        UUID tenantId = CurrentUser.get().tenantId();
        int ditulis = 0;

        for (Object o : lines) {
            if (!(o instanceof Map)) continue;

            Map<String, Object> baris = (Map<String, Object>) o;

            String pesan = Badan.teks(baris, "message");
            if (pesan == null || pesan.isBlank()) continue;

            // Tingkat yang tidak dikenal menjadi INFO, bukan ditolak. Satu
            // salah ketik pada satu baris tidak boleh membuang seluruh kiriman
            // — yang hilang kemudian justru catatan di sekitar kegagalan yang
            // sedang dicari orang.
            String tingkat = Badan.teks(baris, "level", "INFO").toUpperCase(Locale.ROOT);
            if (!TINGKAT_SAH.contains(tingkat)) tingkat = "INFO";

            db.exec("""
                    INSERT INTO logs (tenant_id, level, message, robot_name, machine_name,
                                      process_name, job_id, logged_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, COALESCE(?::timestamptz, now()))
                    """,
                    tenantId, tingkat, pesan,
                    Badan.teks(baris, "robotName"), Badan.teks(baris, "machineName"),
                    Badan.teks(baris, "processName"), Db.uuid(Badan.teks(baris, "jobId")),
                    Badan.teks(baris, "loggedAt"));

            ditulis++;

            // Kesalahan dari robot juga menjadi peringatan. Log dibaca kalau ada
            // yang sengaja mencarinya; peringatan muncul sendiri.
            if ("ERROR".equals(tingkat) || "FATAL".equals(tingkat)) {
                String robot = Badan.teks(baris, "robotName", "robot");
                Peringatan.catat(db, tenantId, "Error", "Kesalahan pada " + robot, pesan, "logs");
            }
        }

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("ok", true);
        hasil.put("written", ditulis);

        return ResponseEntity.ok(hasil);
    }

    /**
     * Buang catatan lama.
     *
     * <p>Batas hari WAJIB dan tidak boleh nol: "hapus semua log" adalah
     * perintah yang tidak bisa dibatalkan, dan bentuk yang paling mudah
     * dijalankan tanpa sengaja.
     */
    @DeleteMapping
    public ResponseEntity<?> bersihkan(@RequestParam(name = "olderThanDays", required = false) Integer hari) {
        if (hari == null || hari < 1) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Parameter 'olderThanDays' wajib diisi dan minimal 1."));
        }

        int terhapus = db.exec(
                "DELETE FROM logs WHERE tenant_id = ? AND logged_at < now() - make_interval(days => ?)",
                CurrentUser.get().tenantId(), hari);

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("ok", true);
        hasil.put("deleted", terhapus);

        return ResponseEntity.ok(hasil);
    }
}
